### ISSUE TITLE
JSX Improvements

### ISSUE THREAD

**Issue by aziz** [bathos/Ecmascript-Sublime#8]
- Comment shortcut inside jsx should insert JSX style comments
https://github.com/babel/babel-sublime/blob/master/Comments%20(JSX).tmPreferences

- babel package highlights the matching tag just like what sublime does on normal HTML/XML files
<img width="224" alt="screen shot 2016-02-10 at 12 31 35 am" src="https://cloud.githubusercontent.com/assets/3202/12934273/adfb81d6-cf8d-11e5-8170-4739a32beacc.png">

- Mark invalid JSX as invalid. just like what you do with js, which is amazing and is much better than any other js package I've ever used.

**Comment by bathos** [comment]
Sounds good. Regarding #3, can you give examples of invalid JSX that currently passes as valid? It can be tough to cover all cases, since the 'invalid' highlighting to some extent is only a side effect of how we match, but if I know what to look for I can probably add it.

**Comment by aziz** [comment]
For example in the screenshot above if I start to delete some characters from `</HEADER>` it should become invalid. additionally `<HEADER ...` should also become invalid since it does not have any closing tag anymore.

**Comment by bathos** [comment]
Hm. As far as I know, there's no mechanism in Sublime Syntax that could allows for dynamic variables like this. If the text were all in one line, it could be done within a single regex pattern, but that won't really help here. So I checked out how BS was doing it, figuring there was maybe a special sublime config for tag matching or something. Instead, I was surprised to find this in the tmLanguage def:

    - begin: (<)([_$a-zA-Z][-$:.\w]*[$\w]*)
      beginCaptures:
        '1': {name: meta.tag.jsx punctuation.definition.tag.begin.jsx}
        '2': {name: meta.tag.jsx entity.name.tag.jsx}
      end: (</)(\2)(>)|(/>)

In other words, tmLanguage did support match reuse like this. This is probably worth asking about on the Sublime forums -- I wasn't aware SL had lost any functionality when comparing to tmLanguage -- or maybe there's a way but it's just not in the docs or something?

**Comment by aziz** [comment]
back references has been supported by tmLanguage and Sublime Syntanx is also support them. 
[from the docs](https://www.sublimetext.com/docs/3/syntax.html):

> ### PHP Heredocs
>This example shows how to match against Heredocs in PHP. The match pattern in the main context captures the heredoc identifier, and the corresponding pop pattern in the heredoc context refers to this captured text with the \1 symbol:
>
>```
name: PHP
scope: source.php
contexts:
  main:
    - match: <<<([A-Za-z][A-Za-z0-9_]*)
      push: heredoc
  heredoc:
    - meta_scope: string.unquoted.heredoc
    - match: ^\1;
        pop: true 
>```

**Comment by bathos** [comment]
Ah, thanks. I was under the impression that backrefs in SS had to live in the same context -- which doesn't actually make sense on reflection -- you're correct, this should work. Thanks!

**Comment by bathos** [comment]
This cannot be done after all, I'm sorry to say. The JSX grammar permits arbitrary whitespace, newlines or comments within the ElementName production, e.g. around the namespace colon and the dot in the member expression production. Therefore we can't actually capture it in a single pattern.

On one hand, the frequency of variations in the wild is probably low, and we could get away with artificially restricting the grammar to achieve this. But it would strike me as very strange to incorrectly mark some grammatical productions as invalid for the sake of marking other semantic productions as invalid (element name matching isn't specified at the grammar level, probably for the very reason that the tokens involved in a match need not be 1:1).

**Comment by aziz** [comment]
I agree. Thanks for all the efforts.
